### DISCUSSION TITLE
GZipDecoder only decodes the first member in gzip data

### DISCUSSION THREAD

**Discussion by lizeyan** [encode/httpx#3269] [Potential Issue] [answered]
### httpx version:
0.27.0

### Current Behavior:
GZipDecoder only decodes the first member in gzip data.

### Expected Behavior:
Be capable of decompressing multi-member gzip data (multiple gzip blocks concatenated together) just like `gzip.decompress`

### Steps To Reproduce:
```
In [1]: import httpx

In [2]: raw_bytes = b'\x1f\x8b\x08\x00\x00\tn\x88\x00\xff\x00\x15\x00\xea\xff{"status": "success",\x03\x00\xeb\xdb\xa3\xb0\x15\x00\x00\x00\x1f\x8
   ...: b\x08\x00\x00\tn\x88\x00\xff\x00\x08\x00\xf7\xff"data": \x03\x00\x1d\xb4\xe6\xc8\x08\x00\x00\x00\x1f\x8b\x08\x00\x00\tn\x88\x00\xff\x00#\
   ...: x00\xdc\xff{"resultType":"matrix","result":[]}\x03\x00\x12\xb7\x95\x1b#\x00\x00\x00\x1f\x8b\x08\x00\x00\tn\x88\x00\xff\x00\x01\x00\xfe\xf
   ...: f}\x03\x00\x0c\xe2\xb6\xfc\x01\x00\x00\x00'

In [3]: from httpx._decoders import GZipDecoder

In [4]: GZipDecoder().decode(raw_bytes)
Out[4]: b'{"status": "success",'

In [5]: import gzip

In [6]: gzip.decompress(raw_bytes)
Out[6]: b'{"status": "success","data": {"resultType":"matrix","result":[]}}'

In [7]:
```

(The raw_bytes are from a Prometheus query request)

### Anything else:
A possible implementation:
```
class GZipDecoder(ContentDecoder):
    ...

    def decode(self, data: bytes) -> bytes:
        decompressed_data = b""
        try:
            length = len(data)
            offset = 0
            while offset < length:
                chunk = self.decompressor.decompress(data[offset:])
                decompressed_data += chunk
                # Update the offset to the next member
                offset += len(data[offset:]) - len(self.decompressor.unused_data)
                if not self.decompressor.unused_data:
                    break
                else:
                    self.decompressor = zlib.decompressobj(zlib.MAX_WBITS | 16)
            return decompressed_data
        except zlib.error as exc:
            raise DecodingError(str(exc)) from exc
```
I have test it and it can handle this case.

**Comment by lizeyan** [comment]
I have opened a pull request: https://github.com/encode/httpx/pull/3270
Seems that this issue has been discovered in 2022? https://github.com/encode/httpx/discussions/2171

**Comment by lovelydinosaur** [comment]
Okay, I don't think we've sufficiently demo'ed that this is an issue yet. (tho it may be.)

Let's ground this against the ecosystem, so it's less abstract...

* Could you demo a public URL that exhibits this behavior?
* Could you check what behavior a couple of browsers have vs. this?
* Could you compare/contrast existing behavior of other Python clients?

**Reply by lizeyan** [reply]
1. The content in my demo is from a private Prometheus. I use Prometheus service from [Volcano Engine](https://www.volcengine.com/product/prometheus)
2. The HTTP standard says that `gzip` refers to [Lempel-Ziv coding](https://en.wikipedia.org/wiki/LZ77_and_LZ78#LZ77) (LZ77) rather than zlib (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding)
3. Here is the implementation of `urllib3` (so as `requests`). It handles multi-member gzip data.
```
class GzipDecoderState:
    FIRST_MEMBER = 0
    OTHER_MEMBERS = 1
    SWALLOW_DATA = 2


class GzipDecoder(ContentDecoder):
    def __init__(self) -> None:
        self._obj = zlib.decompressobj(16 + zlib.MAX_WBITS)
        self._state = GzipDecoderState.FIRST_MEMBER

    def decompress(self, data: bytes) -> bytes:
        ret = bytearray()
        if self._state == GzipDecoderState.SWALLOW_DATA or not data:
            return bytes(ret)
        while True:
            try:
                ret += self._obj.decompress(data)
            except zlib.error:
                previous_state = self._state
                # Ignore data after the first error
                self._state = GzipDecoderState.SWALLOW_DATA
                if previous_state == GzipDecoderState.OTHER_MEMBERS:
                    # Allow trailing garbage acceptable in other gzip clients
                    return bytes(ret)
                raise
            data = self._obj.unused_data
            if not data:
                return bytes(ret)
            self._state = GzipDecoderState.OTHER_MEMBERS
            self._obj = zlib.decompressobj(16 + zlib.MAX_WBITS)

    def flush(self) -> bytes:
        return self._obj.flush()
```

Seems that its implementation is more complete than mine (I did not search it until now)

**Reply by lizeyan** [reply]
Maybe we should just copy their GzipDecoder

**Reply by lovelydinosaur** [reply]
Are you able to check behaviour to see if this is currently supported by Chrome/Firefox/Safari?

From a retrospective... https://github.com/urllib3/urllib3/pull/1289 indicated at that point in time that it was not, though the outlook may have changed since then.

**Reply by lizeyan** [reply]
The use cases for `httpx` and browsers are different. The HTTP requests we query with httpx are often not intended for viewing in a browser.
I cannot access the Prometheus server via browser, but I am sure `curl` and `requests` can decode its response correctly while `httpx` does not.

![img_v3_02dq_dc61c699-7a5b-4537-b15b-ab56d2e045fg](https://github.com/user-attachments/assets/461a302e-1a84-4bec-af90-254b5c6762cd)

![img_v3_02dq_ab2fb2c2-f506-44b4-9cd6-7767eba097cg](https://github.com/user-attachments/assets/a06467d2-6f1b-410c-bdce-177457f0f3f0)

![img_v3_02dq_88b9cf5d-ea8d-4ab0-8f08-a318c37caa2g](https://github.com/user-attachments/assets/ba7f5120-7f23-4c9a-b7b6-f58302c0e629)

This is not a fabricated corner case; it's a real-world scenario.

**Reply by lovelydinosaur** [reply]
@lizeyan Would you be up for taking the example you provided above and creating a runnable example with flask?
That'd be super helpful from my POV as a maintainer, as I can then just focus on confirming and reviewing the PR.

Appreciate your time, thanks.

**Reply by lizeyan** [reply]
```python
from flask import Flask, Response, make_response
import io
import gzip
import requests
import httpx
import threading

app = Flask(__name__)

@app.route('/')
def gzip_response():
    # 创建一个BytesIO对象来存储gzip数据
    buffer = io.BytesIO()

    # 创建一个GzipFile对象,写入多个成员
    with gzip.GzipFile(fileobj=buffer, mode='w') as gz:
        gz.write(b'First member\n')
        gz.close()  # 关闭当前成员

        # 开始新的gzip成员
        gz = gzip.GzipFile(fileobj=buffer, mode='w')
        gz.write(b'Second member\n')
        gz.close()

        # 可以继续添加更多成员...

    # 获取压缩后的数据
    compressed_data = buffer.getvalue()

    # 创建一个Flask响应对象
    response = Response(compressed_data)

    # 设置正确的头部
    response.headers['Content-Type'] = 'text/plain'
    response.headers['Content-Encoding'] = 'gzip'
    response.headers['Content-Length'] = len(compressed_data)

    return response

if __name__ == '__main__':
    threading.Thread(target=lambda: app.run(port=8080), daemon=True).start()
    print(f"requests: {requests.get('http://localhost:8080').text}")
    print(f"httpx   : {httpx.get('http://localhost:8080').text}")
    while True:
        pass
```

It outputs:
```
requests: First member
Second member

httpx   : First member
```


@tomchristie

**Reply by lovelydinosaur** [reply]
Okay yep that replicates.

**Reply by lovelydinosaur** [reply]
What behaviours do Chrome/Firefox/Safari have with the example shown above?

**Reply by lizeyan** [reply, marked as answer]
> What behaviours do Chrome/Firefox/Safari have with the example shown above?

Chrome and Safari only decode the first gzip member.

**Reply by lovelydinosaur** [reply]
Ok, so let's stick with existing behaviour.

 (Thanks for taking the time to work through it)

**Reply by lizeyan** [reply]
It is better to maintain consistent behavior with requests. People won't use httpx to replace browsers, but they will want to use httpx to replace requests.
